Fixed interface name validation checks in tc-root

evershalik opened the pull request:
This PR fixed #25 , which concerns the limitation of tc-root to only allow attachment to interfaces with names containing alphabets or digits, but not hyphens. In the context of hypervisors, tap interfaces commonly contain hyphens, which currently prevents tc-root from attaching to them.

To fix this issue, I have modified the tc-root code to support interface names with hyphens as well. With this change, tc-root will now be able to attach to all interfaces, regardless of their name format.

Please review the changes and let me know if any further modifications are required. Thank you for your attention to this matter.

Signed-off-by: SHANKAR <shankar.s@ramanujan.du.ac.in>

evershalik commented:
so do you want something like this:
```
static int __check_ifname(const char *name)
{
	if (*name == '\0')
		return -1;
	while (*name) {
		if (*name == '/' || isspace(*name))
			return -1;
		++name;
	}
	return 0;
}
```
It means that the name will always be valid except for the condition.
Condition: The name must not be empty, and it must not contain any whitespace or '/' character.
Correct me if I am wrong.

jniesz commented:
> so do you want something like this:
> 
> ```
> static int __check_ifname(const char *name)
> {
> 	if (*name == '\0')
> 		return -1;
> 	while (*name) {
> 		if (*name == '/' || isspace(*name))
> 			return -1;
> 		++name;
> 	}
> 	return 0;
> }
> ```
> 
> It means that the name will always be valid except for the condition. Condition: The name must not be empty, and it must not contain any whitespace or '/' character. Correct me if I am wrong.

Yes, I think it makes sense to align to what the OS supports for naming of interfaces.  @sanfern thoughts?

sanfern commented:
> > so do you want something like this:
> > ```
> > static int __check_ifname(const char *name)
> > {
> > 	if (*name == '\0')
> > 		return -1;
> > 	while (*name) {
> > 		if (*name == '/' || isspace(*name))
> > 			return -1;
> > 		++name;
> > 	}
> > 	return 0;
> > }
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > It means that the name will always be valid except for the condition. Condition: The name must not be empty, and it must not contain any whitespace or '/' character. Correct me if I am wrong.
> 
> Yes, I think it makes sense to align to what the OS supports for naming of interfaces. @sanfern thoughts?

yes, I agree we need to support OS naming standards.

dthaler commented:
> > I think it makes sense to align to what the OS supports for naming of interfaces. @sanfern thoughts?
> 
> yes, I agree we need to support OS naming standards.

Right, so why would we disallow spaces?  On windows the interface friendly name can have spaces.

dthaler commented on tc-root/tc_root_user.c:
```diff
@@ -207,10 +207,14 @@ static bool validate_ifname(const char *input_ifname, char *output_ifname) {
         char c = input_ifname[i];
-
-        if (!(isalpha(c) || isdigit(c)))
+        // Check for each character of input_ifname  if it contains '/' or whitespace characters
+        if (c == '/' || isspace(c))
```
I believe both / and space are legal in interface names on windows

dthaler replied:
As discussed in meeting 4/2, change line 71 to change `dev %s` to `dev '%s'`, and also in dev there it needs to escape any ' to \'

evershalik replied:
Okay I will make changes to the line 71.
but what should I do with interface name validation check? Should we bypass it?

dthaler replied:
The call to `if_nametoindex()` in line 278 is the interface name check.  I am not any validation is needed beyond that.
I am not even sure whether line 211 buys us anything since I would expect an empty string to naturally fail line 278.
So why not just delete the entire `validate_ifname()` function, does it really help with anything?

dthaler replied:
Disagree, unless if_nametoindex() crashes, which it should not, I don't see any value in having additional name or length checks.

evershalik commented:
I have added code to escape any `'` to `\'` for `dev` in line 71  and deleted the `validate_ifname()` function.
Here is the [new commit](https://github.com/l3af-project/eBPF-Package-Repository/pull/42/commits/cb0aac009d005fdc7061388ebeb4d329f653bc7e) with desired changes.
